app/src/main/java/com/ubergeek42/WeechatAndroid/fragments/BufferFragment.java

```diff
@@ -351,14 +360,124 @@ public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
+                    tcViaRelayCompletionResults[tcViaRelayIndex].length()
+                    - tcViaRelayBaseWord.length() + (tcViaRelayAddSpace ? 1 : 0));
+            uiInput.setFocusable(true);
+            uiInput.setFocusableInTouchMode(true);
+            uiInput.setPressed(true);
+            tcViaRelayInProgress = true;
+            return;
+        }
+
+        // in the case of a selection, autocomplete from the left border
+        tcViaRelaySelectionIndex = uiInput.getSelectionStart();
+        tcViaRelayBaseString = uiInput.getText().toString();
+
+        CompletableFuture<RelayObject[]> completionPromise =
+                SendMessageWithPromiseEvent.fire("completion",
+                        "completion 0x%x %d %s", buffer.pointer,
+                        tcViaRelaySelectionIndex, tcViaRelayBaseString);
+
+        // https://weechat.org/files/doc/devel/weechat_relay_protocol.en.html#command_completion
+        new Thread(() -> {
+            try {
+                RelayObject[] objects = completionPromise.get(10, SECONDS);
+                // the documention mentions that "the input string; if not given, completion is
+                // performed on an empty string"
+                // empirically, the Hdata is empty if completion is performed on an empty string
+                if (((Hdata) objects[0]).getCount() > 0) {
+                    HdataEntry item = ((Hdata) objects[0]).getItem(0);
+                    tcViaRelayCompletionResults =
+                            item.getItem("list").asArray().asStringArray();
+                    tcViaRelayAddSpace = item.getItem("add_space").asInt() == 1;
+                    tcViaRelayPosStart = item.getItem("pos_start").asInt();
+                    tcViaRelayPosEnd = item.getItem("pos_end").asInt();
+                    tcViaRelayBaseWord = item.getItem("base_word").asString();
+                    tcViaRelayIndex = 0;
+
+                    if (tcViaRelayCompletionResults.length >= 1) {
+                        StringBuilder st = new StringBuilder(tcViaRelayBaseString);
+                        st.replace(tcViaRelayPosStart, tcViaRelayPosEnd + 1,
+                                tcViaRelayCompletionResults[0]+
+                                        (tcViaRelayAddSpace ? " " : ""));
+
+                        Weechat.runOnMainThread(() -> {
+                            uiInput.setText(st.toString());
+                            uiInput.setSelection(tcViaRelaySelectionIndex +
+                                    tcViaRelayCompletionResults[0].length()
+                                    - tcViaRelayBaseWord.length() + (tcViaRelayAddSpace ? 1 : 0));
+                            if (tcViaRelayCompletionResults.length > 1) {
+                                tcViaRelayInProgress = true;
+                            }
+                        });
+                    } else {
+                        tabCompleteLocalStrategy();
```

Comment:
we are in a worker thread now but `tabCompleteLocalStrategy()` is marked with `@MainThread`

also, perhaps it's best to call it from a single place in a method?